### PR TITLE
IsNullFixer - fix non_yoda_style edge case

### LINKED ISSUE #3071: The is_null fixer breaks on certain files
The is_null fixer worked on every file of our project except for one. It's strange, since similar lines of code (following this `addParam('bleh', (is_null())` format) were successfully converted in other files.

I'm reporting it, since it's apparently a bug in the tokenizer and/or the IsNullFixer.

PHP-CS-Fixer version: 2.6.

Test file which breaks:

- https://raw.githubusercontent.com/mgp25/Instagram-API/b145c9c077d41c599a071629a1b91e13b5aa3242/src/Request/Timeline.php

PHP_CS Configuration:

```php
<?php

return PhpCsFixer\Config::create()
    ->setFinder(
        PhpCsFixer\Finder::create()
        ->in(__DIR__)
    )
    ->setIndent('    ')
    ->setLineEnding("\n")
    ->setRules([
        '@Symfony'                            => true,
        // Override @Symfony rules
        'pre_increment'                       => false,
        'blank_line_before_statement'         => ['statements' => ['return', 'try', 'throw']],
        'phpdoc_align'                        => ['tags' => ['param', 'throws']],
        'method_argument_space'               => ['ensure_fully_multiline' => false],
        'binary_operator_spaces'              => [
            'align_double_arrow' => true,
            'align_equals'       => false,
        ],
        'phpdoc_annotation_without_dot'       => false,
        'yoda_style'                          => [
            // Symfony writes their conditions backwards; we use normal order.
            'equal'            => false,
            'identical'        => false,
            'less_and_greater' => false,
        ],
        'is_null'                             => [
            // Replaces all is_null() with === null.
            'use_yoda_style' => false,
        ],
        // Custom rules
        'align_multiline_comment'             => true,
        'phpdoc_add_missing_param_annotation' => ['only_untyped' => false],
        'ordered_imports'                     => true,
        'phpdoc_order'                        => true,
        'array_syntax'                        => ['syntax' => 'short'],
    ]);
```

Command used:

`./vendor/bin/php-cs-fixer fix --config=.php_cs.dist --allow-risky yes -vvv`

Error output:

```
Loaded config default from ".php_cs.dist".
Using cache file ".php_cs.cache".
SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSESSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS
Legend: ?-unknown, I-invalid file syntax, file ignored, S-Skipped, .-no changes, F-fixed, E-error

Fixed all files in 0.449 seconds, 10.000 MB memory used

Files that were not fixed due to errors reported during fixing:
   1) /home/foo/project/vendor/mgp25/instagram-php/src/Request/Timeline.php

                             
        [ErrorException]     
        Undefined offset: 1  
                             

      {closure}()
        in /home/foo/project/vendor/mgp25/instagram-php/vendor/friendsofphp/php-cs-fixer/src/Tokenizer/Token.php at line 61
      PhpCsFixer\Tokenizer\Token->__construct()
        in /home/foo/project/vendor/mgp25/instagram-php/vendor/friendsofphp/php-cs-fixer/src/Fixer/LanguageConstruct/IsNullFixer.php at line 167
      PhpCsFixer\Fixer\LanguageConstruct\IsNullFixer->applyFix()
        in /home/foo/project/vendor/mgp25/instagram-php/vendor/friendsofphp/php-cs-fixer/src/AbstractFixer.php at line 73
      PhpCsFixer\AbstractFixer->fix()
        in /home/foo/project/vendor/mgp25/instagram-php/vendor/friendsofphp/php-cs-fixer/src/Runner/Runner.php at line 188
      PhpCsFixer\Runner\Runner->fixFile()
        in /home/foo/project/vendor/mgp25/instagram-php/vendor/friendsofphp/php-cs-fixer/src/Runner/Runner.php at line 131
      PhpCsFixer\Runner\Runner->fix()
        in /home/foo/project/vendor/mgp25/instagram-php/vendor/friendsofphp/php-cs-fixer/src/Console/Command/FixCommand.php at line 205
      PhpCsFixer\Console\Command\FixCommand->execute()
        in /home/foo/project/vendor/mgp25/instagram-php/vendor/symfony/console/Command/Command.php at line 264
      [ ... ]
```

### PR DESCRIPTION
**Pull request by keradus**
closes #3071

### CODE REVIEW

**Comment by keradus** [comment]
@SteveJobzniak could you confirm the fix please?

**Comment by keradus** [comment]
(new commit with same content just to retrigger Travis after changing base branch)

**Comment by SteveJobzniak** [comment]
@keradus Okay I downloaded this pull request. Your branch was missing a bunch of 2.6+ features so I removed all fixers except is_null.

Here's my final config:

```php
<?php

return PhpCsFixer\Config::create()
    ->setFinder(
        PhpCsFixer\Finder::create()
        ->in(__DIR__)
    )
    ->setIndent('    ')
    ->setLineEnding("\n")
    ->setRules([
        'is_null'                             => [
            // Replaces all is_null() with === null.
            'use_yoda_style' => false,
        ],
    ]);
```

Test file: https://raw.githubusercontent.com/mgp25/Instagram-API/b145c9c077d41c599a071629a1b91e13b5aa3242/src/Request/Timeline.php

I made a sub-folder of this branch's folder and put the php_cs.conf and Timeline.php there, then I ran:

`../php-cs-fixer fix --config=php_cs.conf --allow-risky yes -vvv`

Result:

```
Loaded config default from "php_cs.conf".
F
Legend: ?-unknown, I-invalid file syntax, file ignored, S-Skipped, .-no changes, F-fixed, E-error
   1) Timeline.php (is_null)

Fixed all files in 0.112 seconds, 8.000 MB memory used
```

And then diffed it compared to the original Timeline.php:

```diff
+ $ diff -u foo.php Timeline.php 
--- foo.php	2017-09-18 00:44:09.000000000 +0200
+++ Timeline.php	2017-09-18 00:43:54.000000000 +0200
@@ -279,8 +279,8 @@
         return $this->ig->request("feed/user/{$userId}/")
             ->addParam('rank_token', $this->ig->rank_token)
             ->addParam('ranked_content', 'true')
-            ->addParam('max_id', (!is_null($maxId) ? $maxId : ''))
-            ->addParam('min_timestamp', (!is_null($minTimestamp) ? $minTimestamp : ''))
+            ->addParam('max_id', ($maxId !== null? $maxId : ''))
+            ->addParam('min_timestamp', ($minTimestamp !== null? $minTimestamp : ''))
             ->getResponse(new Response\UserFeedResponse());
     }
 
@@ -458,6 +458,6 @@
                     touch($filePath, $mediaInfo['taken_at']);
                 }
             }
-        } while (!is_null($nextMaxId = $myTimeline->getNextMaxId()));
+        } while (($nextMaxId = $myTimeline->getNextMaxId()) !== null);
     }
 }
```

It seems to be perfectly correct now?

**Comment by SteveJobzniak** [comment]
Thanks, by the way, for the incredibly fast fixing. Great work!

**Comment by keradus** [comment]
hm... need to see why there is missing space before `?` in first 2 modified lines

**Comment by SteveJobzniak** [comment]
> hm... need to see why there is missing space before "?" in first 2 modified lines

@keradus Sorry that I didn't see your message until now. I could have told you the answer to that. It's a separate bug being fixed here:

https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/3058

File: tests/Fixer/LanguageConstruct/IsNullFixerTest.php [hunk 1]
```diff
@@ -206,4 +195,37 @@ public function provideYodaFixCases()
...
+    }
+
+    public function provideNonYodaFixCases()
+    {
+        return array(
+            array(
+                '<?php $x = $y === null;', '<?php $x = is_null($y);',
+            ),
+            array(
+                '<?php $b = a(a(a(b() === null) === null) === null) === null;',
```
Comment (SteveJobzniak) [inline tests/Fixer/LanguageConstruct/IsNullFixerTest.php]: Wow, this code is beautiful. Thanks for the idea. I think I will start writing all my projects like this. ;)

Comment (keradus) [reply tests/Fixer/LanguageConstruct/IsNullFixerTest.php]: this test is to check if `is_null` inside `is_null` (multiple times) works properly or fixer jump over some of them.
as we don't recommend writing code like that, we don't want to crash if Fixer visit it.

**Comment by SpacePossum** [comment]
Note when merging; this PR https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/3058 touches the same file so might cause conflicts.

File: tests/Fixer/LanguageConstruct/IsNullFixerTest.php (hunk 1 above)
Comment (SteveJobzniak) [reply tests/Fixer/LanguageConstruct/IsNullFixerTest.php]: Yeah I know it's a test, I was kidding. It's good to have that test. I just thought it was the ugliest code style I have ever seen. ;-) If someone does that, then we can really talk about bad coding haha.

**Comment by keradus** [comment]
I merged #3058 and rebase this PR to resolve conflicts

**Comment by SteveJobzniak** [comment]
@keradus Thank you. That's fantastic, I see that both are merged now and that the is_null fixer is finally safe to use (no longer a whitespace killer ;)). Thanks a lot for everything!

**Comment by keradus** [comment]
you welcome ;)
just be aware that they are merged into `@dev` line, they are not yet released with `@stable` channel

**Comment by SteveJobzniak** [comment]
Yep I'll be waiting for 2.6.1. Thanks for everything. :-)

**Pull request closed without merging**